### PR TITLE
zql: remove special-case timebin support

### LINKED ISSUE #828: groupby: remove special-cased time bin handling
Once #535 is complete, we can get rid of the special-case time bin handling in the groupby proc by adding zql sugar that turns `-every X` into a groupby key `ts=Time.trunc(ts, X)`. 

This includes generalizing the "early timebin output" feature that is present in every-groupby, which outputs time-bins as soon as input record time has advanced past a bin. This early-emit will be controlled by a new `-sorted` flag that will tell groupby that the input is sorted on the primary groupby key, and that it is therefore safe to do early-emit.

### LINKED ISSUE #847: groupby "every" assumes forward sorted data
If records presented to a `every X` groupby are not in forward order, then that groupby will emit multiple groups for a given time bin. For example:

```
$ zq -t "every 1h count()" ~/Downloads/sampledata/wrccdc-year1/zng-logs/conn.zng | head -n 10
#0:record[ts:time,count:uint64]
0:[1490382000;88795;]
0:[1490382000;100;]
0:[1490382000;100;]
0:[1490382000;100;]
0:[1490382000;92;]
0:[1490382000;98;]
0:[1490382000;100;]
0:[1490382000;92;]
0:[1490382000;100;]
```

### PR DESCRIPTION
**Pull request by henridf**
This PR removes the special-case "timebinning" support in the groupby proc, using the new computed groupby key facility introduced in #860. The zql `every` syntax remains such that `every x count() by y` is effectively translated to `count() by ts=time.Trunc(ts, x), y`.

The prior timebinning assumed that input was always forward-sorted in time, and it would output time-bin keys as soon as time (as observed on input records) had advanced past a bin. This assumption is a remnant of a prior system (when we believed that "time was on our side") where records _were_ guaranteed to be time-sorted, but the guarantee is not present in zq or zqd, and so the assumption could lead to silently incorrect results.

Nonetheless, the early emit facility is useful, and has been reimplemented, in a general way that is no specific to time bins. It is enabled by passing the `-sorted <dir>` argument to groupby (where `dir` should be `1` or `-1` to indicate the input sort order), which signals that the input is sorted in the primary key.

For example: `every 1h count() by _path -sorted 1` could be used to tell groupby that input records are forward-ordered by time and can be early-emitted.

closes #828 
closes #847

### CODE REVIEW

**Comment by henridf** [comment]
This still needs a companion Brim PR before merging... but should be otherwise ready for review.

**Comment by henridf** [comment]
The companion brim PR is https://github.com/brimsec/brim/pull/864

File: ast/ast.go [hunk 1]
```diff
@@ -263,12 +263,15 @@ type (
 	// large time ranges are processed and streamed efficiently.
 	// The limit parameter specifies the number of different groups that can
 	// be aggregated over. When absent, the runtime defaults to an appropriate value.
+	//
+	// xxx comment sorted
```
Comment (nwt) [inline ast/ast.go:267]: Fix me!

File: ast/ast.go [hunk 2]
```diff
@@ -263,12 +263,15 @@ type (
...
 	GroupByProc struct {
 		Node
 		Duration Duration     `json:"duration"`
 		Limit    int          `json:"limit,omitempty"`
 		Keys     []Assignment `json:"keys"`
 		Reducers []Reducer    `json:"reducers"`
+		Sorted   int          `json:"sorted"`
```
Comment (nwt) [inline ast/ast.go:274]:
```suggestion
		Sorted   int          `json:"sorted,omitempty"`
```

File: proc/groupby.go [hunk 3]
```diff
@@ -381,26 +396,29 @@ func (g *GroupByAggregator) Results(eof bool, minTs nano.Ts, maxTs nano.Ts) (zbu
...
 	for _, k := range keys {
-		row := table[k]
-		var zv zcode.Bytes
-		if g.TimeBinDuration > 0 {
-			zv = zcode.AppendPrimitive(zv, zng.EncodeTime(row.ts))
+		row := g.table[k]
+		if !eof {
+			if g.valueSortFn(*row.primary, *g.maxKey) >= 0 {
+				continue
+			}
```
Comment (nwt) [inline proc/groupby.go:418]:
```suggestion
		if !eof && g.valueSortFn(*row.primary, *g.maxKey) >= 0 {
			continue
```

File: proc/groupby.go [hunk 4]
```diff
@@ -316,62 +353,40 @@ func (g *GroupByAggregator) Consume(r *zng.Record) error {
...
-// times; all completed time bins at the time of the invocation are
-// returned. A final call with eof=true should be made to get the
-// final (possibly incomplete) time bin.
-// If this is not a time-binned aggregation, a single call (with
-// eof=true) should be made after all records have been Consumed()'d.
-func (g *GroupByAggregator) Results(eof bool, minTs nano.Ts, maxTs nano.Ts) (zbuf.Batch, error) {
-	var bins []nano.Ts
-	for b := range g.tables {
-		bins = append(bins, b)
+func (g *GroupByAggregator) updateMax(v zng.Value) {
```
Comment (nwt) [inline proc/groupby.go:368]: Nit: updateMaxKey might be a better name.

File: proc/groupby.go [hunk 5]
```diff
@@ -207,26 +237,27 @@ func (g *GroupBy) Pull() (zbuf.Batch, error) {
 		}
 	}
 	batch.Unref()
-	if g.timeBinned {
-		f, err := g.agg.Results(false, g.MinTs, g.MaxTs)
+	if g.agg.inputSortDir != 0 {
+		arr, err := g.agg.Results(false)
```
Comment (nwt) [inline proc/groupby.go:241]: Nit: Call this batch.

File: ast/ast.go (hunk 2 above)
Comment (nwt) [reply ast/ast.go:274]: Also, Sorted is a pretty vague name, so I think we should call this InputSortDir/input_sort_dir.

File: proc/groupby.go [hunk 6]
```diff
@@ -25,11 +25,11 @@ type GroupByKey struct {
...
-	duration ast.Duration
-	limit    int
-	keys     []GroupByKey
-	reducers []compile.CompiledReducer
-	builder  *ColumnBuilder
+	limit        int
+	keys         []GroupByKey
+	reducers     []compile.CompiledReducer
+	builder      *ColumnBuilder
+	inputSortDir int
```
Comment (nwt) [inline proc/groupby.go:32]: Keep the configuration fields together and (mostly) sorted.
```suggestion
	inputSortDir int
	limit        int
	keys         []GroupByKey
	reducers     []compile.CompiledReducer
	builder      *ColumnBuilder
```

File: proc/groupby.go [hunk 7]
```diff
@@ -47,8 +47,32 @@ const defaultGroupByLimit = 1000000
...
+					&ast.Literal{ast.Node{"Literal"}, "int64", strconv.Itoa(node.Duration.Seconds)},
+				},
+			},
+		}
+		astKeys = append([]ast.Assignment{everyKey}, astKeys...)
+	}
+	if len(astKeys) == 0 {
+		return nil, fmt.Errorf("compiling groupby: must have at least one key (or -every)")
+	}
+	for _, astKey := range astKeys {
```
Comment (nwt) [inline proc/groupby.go:75]: Please don't add 25 lines of unrelated code between the declarations of keys and targetNames and their uses.

File: ast/ast.go (hunk 2 above)
Comment (henridf) [reply ast/ast.go:274]: 👍

Comment (henridf) [reply ast/ast.go:274]: More importantly, do you think the zql flag `-sorted` should also be renamed?

File: ast/ast.go (hunk 1 above)
Comment (henridf) [reply ast/ast.go:267]: Fixed

File: proc/groupby.go (hunk 3 above)
Comment (henridf) [reply proc/groupby.go:418]: oops

File: proc/utils.go [hunk 8]
```diff
@@ -197,7 +197,7 @@ func (p *ProcTest) Finish() error {
 	}
 }
 
-func parse(zctx *resolver.Context, src string) (*zbuf.Array, error) {
+func ParseTestZng(zctx *resolver.Context, src string) (*zbuf.Array, error) {
```
Comment (nwt) [inline proc/utils.go:200]: Nit: This parses tzng, not zng, so ParseTzng might be a better name.

File: ast/ast.go (hunk 2 above)
Comment (nwt) [reply ast/ast.go]: I don't think we should expose this through ZQL.

Instead, I think we should add a `sortKey string` parameter to driver.Compile*, set it appropriately when calling those functions (to `"ts"` in zqd; to `""` in zq for now, ultimately wiring it to a command line flag), and then use it as input to an analysis that will set ast.GroupByProc.InputSortDir to the correct value before calling proc.CompileProc.

That's obviously more work, but we don't need to do a thorough job of it now. We just need enough to catch the Brim histogram AST, which seems pretty straightforward.

File: ast/ast.go [hunk 9]
```diff
@@ -255,20 +255,25 @@ type (
...
+	// duration field is non-zero, then the groups further partioned by time
+	// into bins of the duration. In this case time is the primary key.
+	// The Limit field specifies the number of different groups that can be
+	// aggregated over. When absent, the runtime defaults to an
+	// appropriate value.
+	// The InputSortDir field indicates that input is sorted (with
+	// direction indicated by the sign of the field) in the
+	// primary key. In this case, the proc outputs the reducer
+	// results from each key as they complete so that large inputs
+	// are processed and streamed efficiently.
```
Comment (nwt) [inline ast/ast.go:269]:
```suggestion
	// of the fields specified in the keys field (where the first key is the
	// grouping key), and applies reducers (if any) to each group.  If the
	// Duration field is non-zero, then the groups are further partitioned by time
	// into bins of the duration. In this case, the grouping key is always ts.
	// The Limit field specifies the number of different groups that can be
	// aggregated over. When absent, the runtime defaults to an
	// appropriate value.
	// The InputSortDir field indicates that input is sorted (with
	// direction indicated by the sign of the field) in the
	// grouping key. In this case, the proc outputs the reducer
	// results from each key as they complete so that large inputs
	// are processed and streamed efficiently.
```

File: proc/groupby.go [hunk 10]
```diff
@@ -46,9 +46,33 @@ func IsErrTooBig(err error) bool {
 const defaultGroupByLimit = 1000000
 
 func CompileGroupBy(node *ast.GroupByProc, zctx *resolver.Context) (*GroupByParams, error) {
+	astKeys := node.Keys
+
+	if node.InputSortDir != 0 && node.Duration.Seconds == 0 && len(astKeys) > 0 {
+		if _, ok := astKeys[0].Expr.(ast.FieldExpr); !ok {
+			return nil, fmt.Errorf("compiling groupby: cannot use -sorted in conjunction with a computed primary key")
```
Comment (nwt) [inline proc/groupby.go:53]:
```suggestion
			return nil, fmt.Errorf("compiling groupby: cannot use -sorted in conjunction with a computed grouping key")
```

File: proc/groupby.go [hunk 11]
```diff
@@ -131,63 +151,73 @@ type GroupByAggregator struct {
...
-	limit           int
+	kctx         *resolver.Context
+	keys         []GroupByKey
+	reducerDefs  []compile.CompiledReducer
+	builder      *ColumnBuilder
+	table        map[string]*GroupByRow
+	reverse      bool
+	logger       *zap.Logger
+	limit        int
+	valueSortFn  expr.ValueSortFn // to compare primary keys for early key output
```
Comment (nwt) [inline proc/groupby.go:162]:
```suggestion
	valueSortFn  expr.ValueSortFn // to compare grouping keys for early key output
```

File: proc/groupby.go [hunk 12]
```diff
@@ -316,62 +353,40 @@ func (g *GroupByAggregator) Consume(r *zng.Record) error {
...
-		newRecs, err := g.recordsForTable(g.tables[b])
-		if err != nil {
-			return nil, err
-		}
-		recs = append(recs, newRecs...)
-		delete(g.tables, b)
+}
+
+// Results returns a batch of aggregation result records. If the input
+// is sorted in the primary key, this can be called multiple times;
```
Comment (nwt) [inline proc/groupby.go:379]:
```suggestion
// is sorted in the grouping key, this can be called multiple times;
```

File: proc/groupby.go [hunk 13]
```diff
@@ -131,63 +151,73 @@ type GroupByAggregator struct {
...
+	recordSortFn expr.SortFn
+	maxKey       *zng.Value
+	inputSortDir int
 }
 
 type GroupByRow struct {
 	keycols  []zng.Column
 	keyvals  zcode.Bytes
-	ts       nano.Ts
+	primary  *zng.Value // for sorting when input sorted
```
Comment (nwt) [inline proc/groupby.go:171]:
```suggestion
	group  *zng.Value // for sorting when input sorted
```
Or groupval or groupVal or something else.

File: proc/groupby.go [hunk 14]
```diff
@@ -46,9 +46,33 @@ func IsErrTooBig(err error) bool {
...
+					&ast.FieldRead{ast.Node{"FieldRead"}, "ts"},
+					&ast.Literal{ast.Node{"Literal"}, "int64", strconv.Itoa(node.Duration.Seconds)},
+				},
+			},
+		}
+		astKeys = append([]ast.Assignment{everyKey}, astKeys...)
+	}
+	if len(astKeys) == 0 {
+		return nil, fmt.Errorf("compiling groupby: must have at least one key (or -every)")
+	}
```
Comment (nwt) [inline proc/groupby.go:72]: I think it would be simpler to do this transformation in zql.peg instead of here. That'd let us drop ast.GroupByProc.Duration, too.

File: ast/ast.go (hunk 9 above)
Comment (henridf) [reply ast/ast.go:269]: Applied but defined first key as "primary grouping key" rather than "grouping key" because all keys are grouping keys.

File: proc/groupby.go (hunk 14 above)
Comment (henridf) [reply proc/groupby.go:72]: (You mean in parser-support.go and parser-support.js, right?)

I had tried that early on but it wasn't simpler. More importantly, I don't think that parser code should be aware of specific zql functions like "Time.trunc", or doing this kind of transformation.  Ideally it would go in some intermediate compilation pass, but we just have one for the time being... so I put it there.

Comment (henridf) [reply proc/groupby.go:72]: (separately, and independently of this PR, Duration should be called Every).
